### ISSUE TITLE
Allow `swiftlint -h` and `swiftlint --help` for all subcommands

### ISSUE THREAD

**Issue by AliSoftware** [realm/SwiftLint#531]
I for one keep trying `swiftlint -h` and `swiftlint --help` before realizing the expected syntax is `swift help`

Also, using the `-h` / `--help` flags is more POSIX compliant.

One advantage of this syntax is that you usually start typing the command (like `swift lint`) and in the middle of it realize you are not sure about the possible parameters, so you can add that `-h` at the end of the command line to check them, without having to move the cursor left to add the `help` work at the beginning of the command line.
Then one you've read the help, you can just use the up-arrow to recall the last command line typed, delete that `-h` flag and continue typing where you stopped your unfinished command.

**Comment by mr-fixit** [comment]
@AliSoftware, this feature you mention, being able to "add -h to the end of the command line", does that work even if the command line up to there isn't valid?  Or does it have to be a valid command up to that point, so that `swiftlint <foo> --help` would only be valid if `swiftlint help <foo>` were valid?

**Comment by AliSoftware** [comment]
🤔  that's a very good question. I guess I don't care, I mean whatever is the easiest to implement would be fine to me.

I think what would make more sense tho would be for the command to be valid, because `swiftlint foo -h` would do the same as `swiftlint help foo`. But I guess it also depends on what you call "valid": my point is both `swiftlint foo -h` and `swiftlint help foo` should behave the same, and that behavior is probably just to print an error message telling that there subcommand `foo` doesn't exist then show the usage.

What would make more sense would actually to be POSIX compliant and behave the same as other common UNIX tools regarding this `-h` behavior. See CocoaPods and its `pod foo -h` behavior for example.

**Comment by orta** [comment]
Yeah, plus one, I got real confused about how to even use the tool as a first time user

<img width="882" alt="screen shot 2017-10-24 at 9 03 18 am" src="https://user-images.githubusercontent.com/49038/31944470-3680f508-b89a-11e7-9f02-885636a6c355.png">

> this feature you mention, being able to "add -h to the end of the command line", does that work even if the command line up to there isn't valid? 

Yep, `-h` or `--help` should ignore any command and give the information about the subcommand

**Comment by jpsim** [comment]
For the record, I agree with this and would like to move away from Commandant to a CLI framework that's simple, easy to use and type-safe. Maybe that's [Commander](https://github.com/kylef/Commander)? Maybe that's something else. But I don't want to hack this in by working around Commandant's limitations.

**Comment by AliSoftware** [comment]
FYI `Commander` is what we use in SwiftGen and it already supports that feature.
Not saying that's the only solution and that's what SwiftLint should choose (it might have some other limitations not suitable for SwiftLint?? I'll let you check depending on your needs), but at least there wouldn't be anything to hack for that "--help" feature.

**Comment by salutis** [comment]
> Please comment to prevent this issue from being closed.

This is still an issue.

(These “agile” bots are a stupid idea.)

**Comment by jpsim** [comment]
We've migrated to [swift-argument-parser](https://github.com/apple/swift-argument-parser), which handles this better than Commandant did.

```shell
$ # Before
$ swiftlint --help
Unrecognized arguments: --help
$ # After
$ swiftlint --help
OVERVIEW: A tool to enforce Swift style and conventions.

USAGE: swiftlint <subcommand>

OPTIONS:
  --version               Show the version.
  -h, --help              Show help information.

SUBCOMMANDS:
  analyze                 Run analysis rules
  docs                    Open SwiftLint documentation website in the default web browser
  generate-docs           Generates markdown documentation for all rules
  lint (default)          Print lint warnings and errors
  rules                   Display the list of rules and their identifiers
  version                 Display the current version of SwiftLint

  See 'swiftlint help <subcommand>' for detailed help.
```

I'd appreciate if you could try out SwiftLint on the `master` branch and share if anything's improved or regressed for you.
